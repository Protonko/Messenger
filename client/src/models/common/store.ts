export interface IAction<T extends string, U = undefined, V = undefined> {
  type: T
  payload: U
  meta: V
}

export interface IBuildReducer<T, U, V extends string, M = undefined> {
  (state: T, action: IAction<V, U, M>): T
}

export type TReducers<T, U, V extends string, M = undefined> = Record<string, IBuildReducer<T, U, V, M>>
