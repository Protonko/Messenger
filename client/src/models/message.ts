import type {TStatusResponse} from './common/fetch'

export interface ICreateMessageBody {
  id: string
  text: string
}

export interface ICreateMessageResponse {
  status: TStatusResponse
  read: boolean
  // TODO: FIX
  attachments: Array<any>
  id: string
  user: string
  createdAt: string
  updatedAt: string
}

export interface IMessage {
  read: boolean
  // TODO: FIX
  attachments: Array<any>
  id: string
  user: string
  createdAt: string
  updatedAt: string
}
