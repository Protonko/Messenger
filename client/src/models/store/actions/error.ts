import type {AnyAction} from 'redux'

export enum ErrorActionTypes {
  INVALID_TOKEN_ERROR = '[ERROR]INVALID_TOKEN_ERROR',
  HIDE_ERROR_NOTIFICATION = '[ERROR]HIDE_ERROR_NOTIFICATION',
  COMMON_ERROR = '[ERROR]COMMON_ERROR_ACTION',
}

export interface InvalidTokenErrorAction extends AnyAction {
  type: ErrorActionTypes.INVALID_TOKEN_ERROR
  payload: string
}

export interface CommonErrorAction extends AnyAction {
  type: ErrorActionTypes.COMMON_ERROR
  payload: string
}

export interface HideErrorNotificationAction extends AnyAction {
  type: ErrorActionTypes.HIDE_ERROR_NOTIFICATION
}

export type AllErrorActions =
  | InvalidTokenErrorAction
  | CommonErrorAction
  | HideErrorNotificationAction
