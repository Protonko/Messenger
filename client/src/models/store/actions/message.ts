import type {AnyAction} from 'redux'
import type {IMessage} from 'models/message'

export enum MessageActionsTypes {
  CREATE_MESSAGE = '[MESSAGE]CREATE_MESSAGE',
  CREATE_MESSAGE_SUCCESS = '[MESSAGE]CREATE_MESSAGE_SUCCESS',
  CREATE_MESSAGE_ERROR = '[MESSAGE]CREATE_MESSAGE_ERROR',
  GET_MESSAGES = '[MESSAGE]GET_MESSAGES',
  GET_MESSAGES_SUCCESS = '[MESSAGE]GET_MESSAGES_SUCCESS',
  GET_MESSAGES_ERROR = '[MESSAGE]GET_MESSAGES_ERROR',
  APPEND_MESSAGE = '[MESSAGE]APPEND_MESSAGE',
}

// payloads
export interface IGetMessagesSuccessPayload {
  messages: IMessage[]
  dialogId: string
}

export interface IGetMessagesErrorPayload {
  errorMessage: string
  dialogId: string
}

export interface ICreateMessagePayload {
  text: string
  dialogId: string
  interlocutorId: string
}

// actions
export interface CreateMessageAction extends AnyAction {
  type: MessageActionsTypes.CREATE_MESSAGE
  payload: ICreateMessagePayload
}

export interface CreateMessageSuccessAction extends AnyAction {
  type: MessageActionsTypes.CREATE_MESSAGE_SUCCESS
  payload: IMessage
}

export interface CreateMessageErrorAction extends AnyAction {
  type: MessageActionsTypes.CREATE_MESSAGE_ERROR
  payload: string
}

export interface GetMessagesAction extends AnyAction {
  type: MessageActionsTypes.GET_MESSAGES
  payload: string
}

export interface GetMessagesSuccessAction extends AnyAction {
  type: MessageActionsTypes.GET_MESSAGES_SUCCESS
  payload: IGetMessagesSuccessPayload
}

export interface GetMessagesErrorAction extends AnyAction {
  type: MessageActionsTypes.GET_MESSAGES_ERROR
  payload: IGetMessagesErrorPayload
}

export interface AppendMessageAction extends AnyAction {
  type: MessageActionsTypes.APPEND_MESSAGE
  payload: IMessage
}

export type AllMessageActions =
  | CreateMessageAction
  | CreateMessageSuccessAction
  | CreateMessageErrorAction
  | GetMessagesAction
  | GetMessagesSuccessAction
  | GetMessagesErrorAction
  | AppendMessageAction
