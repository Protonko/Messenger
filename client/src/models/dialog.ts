import type {IUser} from './user'
import {Status, ReadStatus} from './common/status'

export interface IDialog {
  id: string
  interlocutor: IUser,
  lastMessage: string
  createdAt: string
  updatedAt: string
  messages: number
  status: Status
  readStatus: ReadStatus | null
}

export interface ICreateDialogBody {
  author: string
  interlocutor: string
  text: string
}
