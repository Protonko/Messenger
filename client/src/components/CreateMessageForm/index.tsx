import type {EmojiData} from 'emoji-mart'
import type {RootState} from 'store/reducers'
import {useState, useCallback, useMemo, FormEvent} from 'react'
import {useDispatch, useSelector} from 'react-redux'
import {ReactComponent as Clip} from 'assets/icons/clip.svg'
import {ReactComponent as Microphone} from 'assets/icons/microphone.svg'
import {EVENTS_SOCKET} from 'models/common/socket'
import {socket} from 'utils/socket'
import throttle from 'utils/throttle'
import {useSearchParams} from 'hooks/useSearchParams'
import {TYPING_TIMEOUT} from 'static/constants'
import {createMessage} from 'store/actions/message'
import {commonError} from 'store/actions/error'
import {Textarea} from 'components/common/Textarea'
import {Avatar} from 'components/common/Avatar'
import {Button} from 'components/common/Button'
import {Emoji} from 'components/common/Emoji'

export const CreateMessageForm = () => {
  const dialogParam = useSearchParams('dialog')
  const {dialogs, account} = useSelector((state: RootState) => ({
    ...state.dialogs,
    ...state.auth,
  }))
  const dispatch = useDispatch()
  const [value, setValue] = useState('')
  const interlocutor = useMemo(
    () => dialogs?.find((dialog) => dialog.id === dialogParam)?.interlocutor,
    [dialogs, dialogParam],
  )

  const emitWriteMessage = useCallback(
    throttle(() => {
      interlocutor && socket.emit(EVENTS_SOCKET.TYPING_MESSAGE, interlocutor.id)
    }, TYPING_TIMEOUT),
    [dialogs],
  )

  const onChange = (value: string) => {
    emitWriteMessage()
    setValue(value)
  }

  const onSubmit = (event: FormEvent<HTMLFormElement>) => {
    event.preventDefault()

    if (!interlocutor || !dialogParam) {
      return dispatch(commonError('ID not found!'))
    }

    dispatch(
      createMessage({
        text: value,
        dialogId: dialogParam,
        interlocutorId: interlocutor.id,
      }),
    )
  }

  const onSelect = (data: EmojiData) => {
    if ('native' in data) {
      setValue((prev) => `${prev} ${data.native}`)
    }
  }

  return (
    <form className="create-message-form" onSubmit={onSubmit}>
      <div className="create-message-form__row">
        <Avatar name={account?.full_name} src={account?.avatar ?? ''} />
        <Textarea
          placeholder="Write a message..."
          className="create-message-form__textarea"
          value={value}
          onChange={onChange}
        />
        <Avatar
          name={interlocutor?.full_name}
          src={interlocutor?.avatar ?? ''}
        />
      </div>

      <div className="create-message-form__row create-message-form__actions">
        <div className="create-message-form__col">
          <ul className="create-message-form__icons list list--reset">
            <li className="create-message-form__icons-item">
              <Clip className="create-message-form__icon" />
            </li>
            <li className="create-message-form__icons-item">
              <Microphone className="create-message-form__icon" />
            </li>
            <li className="create-message-form__icons-item">
              <Emoji onSelect={onSelect} />
            </li>
          </ul>
        </div>

        <div className="create-message-form__col">
          <Button type="submit" text="SEND" />
        </div>
      </div>
    </form>
  )
}
