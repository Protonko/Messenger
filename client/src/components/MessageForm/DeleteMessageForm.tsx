import {useContext} from 'react'
import {useDispatch} from 'react-redux'
import {Button, ButtonModifier} from 'components/common/Button'
import {ChatContext} from 'context/ChatContext'
import {deleteMessages} from 'store/actions/message'
import {useSearchParams} from 'hooks/useSearchParams'

export const DeleteMessageForm = () => {
  const dispatch = useDispatch()
  const dialogId = useSearchParams('dialog')
  const {resetSelectedMessagesIds, selectedMessagesIds} =
    useContext(ChatContext)

  const onDelete = () => {
    if (dialogId) {
      dispatch(deleteMessages({messagesIds: selectedMessagesIds, dialogId}))
      resetSelectedMessagesIds()
    }
  }

  return (
    <div className="delete-message-form">
      <Button
        additionalClassName="delete-message-form__button delete-message-form__button--delete"
        text="DELETE"
        onClick={onDelete}
      />
      <Button
        additionalClassName="delete-message-form__button delete-message-form__button--cancel"
        modifier={ButtonModifier.LINKING}
        text="CANCEL"
        onClick={resetSelectedMessagesIds}
      />
    </div>
  )
}
