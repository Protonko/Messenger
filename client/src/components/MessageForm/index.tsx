import {useContext} from 'react'
import {ChatContext} from 'context/ChatContext'
import {DeleteMessageForm} from './DeleteMessageForm'
import {CreateMessageForm} from './CreateMessageForm'

export const MessageForm = () => {
  const {selectedMessagesIds} = useContext(ChatContext)

  if (selectedMessagesIds.length) {
    return <DeleteMessageForm />
  }

  return <CreateMessageForm />
}
