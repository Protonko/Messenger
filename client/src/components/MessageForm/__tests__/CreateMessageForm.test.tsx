import type {MockStoreEnhanced} from 'redux-mock-store'
import type {RootState} from 'store/reducers'
import {mount, ReactWrapper} from 'enzyme'
import {Provider} from 'react-redux'
import {Router} from 'react-router-dom'
import {createMemoryHistory, History, Location} from 'history'
import {ErrorActionTypes} from 'models/store/actions/error'
import {MessageActionsTypes} from 'models/store/actions/message'
import {CreateMessageForm} from 'components/MessageForm/CreateMessageForm'
import {createMockStore, state} from 'static/test-mocks/store'
import {DIALOG} from 'static/test-mocks/dialog'

const ELEMENT_SELECTORS = {
  form: 'create-message-form',
  textarea: 'textarea__form',
}

describe('CreateMessageForm', () => {
  let component: (store: MockStoreEnhanced<RootState>) => ReactWrapper
  let history: History
  let stateRedux: RootState

  const historyLocation: Location = {
    pathname: '',
    search: `dialog=${DIALOG.id}`,
    hash: '',
    key: '',
    state: null,
  }

  beforeEach(() => {
    stateRedux = {...state}
    history = {
      ...createMemoryHistory(),
      push: jest.fn(),
      goBack: jest.fn(),
      go: jest.fn(),
      location: historyLocation,
    }

    component = (store) =>
      mount(
        <Provider store={store}>
          <Router history={history}>
            <CreateMessageForm />
          </Router>
        </Provider>,
      )
  })

  it('Should render component', () => {
    expect(component(createMockStore(stateRedux))).toMatchSnapshot(
      'CreateMessageForm',
    )
  })

  it('Should change input value', () => {
    const componentMount = component(createMockStore(stateRedux))
    componentMount
      .find(`.${ELEMENT_SELECTORS.textarea}`)
      .simulate('change', {target: {value: 'test'}})
    expect(
      componentMount.find(`.${ELEMENT_SELECTORS.textarea}`).prop('value'),
    ).toBe('test')
  })

  it('Should submit form', () => {
    stateRedux = {
      ...stateRedux,
      dialogs: {
        ...stateRedux.dialogs,
        dialogs: [DIALOG],
      },
    }

    const payload = {
      attachment: undefined,
      dialogId: 'id',
      interlocutorId: '123',
      text: '',
    }

    const store = createMockStore(stateRedux)
    store.dispatch = jest.fn()
    const componentMount = component(store)
    componentMount.find(`.${ELEMENT_SELECTORS.form}`).simulate('submit')
    expect(store.dispatch).toBeCalledWith({
      type: MessageActionsTypes.CREATE_MESSAGE,
      payload,
    })
  })

  it('Should submit form with error', () => {
    const store = createMockStore(stateRedux)
    store.dispatch = jest.fn()
    const componentMount = component(store)
    componentMount.find(`.${ELEMENT_SELECTORS.form}`).simulate('submit')
    expect(store.dispatch).toBeCalledWith({
      type: ErrorActionTypes.COMMON_ERROR,
      payload: 'ID not found!',
    })
  })
})
