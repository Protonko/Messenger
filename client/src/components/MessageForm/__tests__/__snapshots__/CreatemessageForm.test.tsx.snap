// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CreateMessageForm Should render component: CreateMessageForm 1`] = `
<Provider
  store={
    Object {
      "clearActions": [Function],
      "dispatch": [Function],
      "getActions": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
>
  <Router
    history={
      Object {
        "action": "POP",
        "block": [Function],
        "canGo": [Function],
        "createHref": [Function],
        "entries": Array [
          Object {
            "hash": "",
            "key": "i",
            "pathname": "/",
            "search": "",
            "state": undefined,
          },
        ],
        "go": [MockFunction],
        "goBack": [MockFunction],
        "goForward": [Function],
        "index": 0,
        "length": 1,
        "listen": [Function],
        "location": Object {
          "hash": "",
          "key": "",
          "pathname": "",
          "search": "dialog=id",
          "state": null,
        },
        "push": [MockFunction],
        "replace": [Function],
      }
    }
  >
    <CreateMessageForm>
      <form
        className="create-message-form"
        onSubmit={[Function]}
      >
        <div
          className="create-message-form__row"
        >
          <Avatar
            src=""
          >
            <div
              className="avatar"
            >
              <img
                alt=""
                className="avatar__image"
                src=""
              />
              <Loader>
                <div
                  className="loader"
                >
                  <div
                    className="loader__line"
                    key="0"
                  />
                  <div
                    className="loader__line"
                    key="1"
                  />
                  <div
                    className="loader__line"
                    key="2"
                  />
                  <div
                    className="loader__line"
                    key="3"
                  />
                  <div
                    className="loader__line"
                    key="4"
                  />
                  <div
                    className="loader__line"
                    key="5"
                  />
                  <div
                    className="loader__line"
                    key="6"
                  />
                  <div
                    className="loader__line"
                    key="7"
                  />
                  <div
                    className="loader__line"
                    key="8"
                  />
                  <div
                    className="loader__line"
                    key="9"
                  />
                  <div
                    className="loader__line"
                    key="10"
                  />
                  <div
                    className="loader__line"
                    key="11"
                  />
                </div>
              </Loader>
            </div>
          </Avatar>
          <div
            className="create-message-form__fields"
          >
            <Textarea
              className="create-message-form__textarea"
              onChange={[Function]}
              placeholder="Write a message..."
              value=""
            >
              <div
                className="textarea create-message-form__textarea"
              >
                <textarea
                  className="textarea__form"
                  onChange={[Function]}
                  placeholder="Write a message..."
                  value=""
                />
              </div>
            </Textarea>
          </div>
          <Avatar
            src=""
          >
            <div
              className="avatar"
            >
              <img
                alt=""
                className="avatar__image"
                src=""
              />
              <Loader>
                <div
                  className="loader"
                >
                  <div
                    className="loader__line"
                    key="0"
                  />
                  <div
                    className="loader__line"
                    key="1"
                  />
                  <div
                    className="loader__line"
                    key="2"
                  />
                  <div
                    className="loader__line"
                    key="3"
                  />
                  <div
                    className="loader__line"
                    key="4"
                  />
                  <div
                    className="loader__line"
                    key="5"
                  />
                  <div
                    className="loader__line"
                    key="6"
                  />
                  <div
                    className="loader__line"
                    key="7"
                  />
                  <div
                    className="loader__line"
                    key="8"
                  />
                  <div
                    className="loader__line"
                    key="9"
                  />
                  <div
                    className="loader__line"
                    key="10"
                  />
                  <div
                    className="loader__line"
                    key="11"
                  />
                </div>
              </Loader>
            </div>
          </Avatar>
        </div>
        <div
          className="create-message-form__row create-message-form__actions"
        >
          <div
            className="create-message-form__col"
          >
            <ul
              className="create-message-form__icons list list--reset"
            >
              <li
                className="create-message-form__icons-item"
              >
                <FileUploader
                  onChange={[Function]}
                >
                  <div
                    className="file-uploader"
                  >
                    <label
                      className="file-uploader__label"
                    >
                      <div
                        className="file-uploader__icon"
                      />
                      <input
                        className="file-uploader__input"
                        name="file"
                        onChange={[Function]}
                        type="file"
                      />
                    </label>
                  </div>
                </FileUploader>
              </li>
              <li
                className="create-message-form__icons-item"
              >
                <Emoji
                  onSelect={[Function]}
                >
                  <div
                    className="emoji"
                  >
                    <div
                      className="emoji__button"
                      onClick={[Function]}
                    />
                  </div>
                </Emoji>
              </li>
            </ul>
          </div>
          <div
            className="create-message-form__col"
          >
            <Button
              text="SEND"
              type="submit"
            >
              <button
                className="button"
                type="submit"
              >
                <span
                  className="button__text"
                >
                  SEND
                </span>
              </button>
            </Button>
          </div>
        </div>
      </form>
    </CreateMessageForm>
  </Router>
</Provider>
`;
